### ISSUE TITLE
Shouldn't `split column` and `split chars` be `str` subcommands?

### ISSUE THREAD

**Issue by webbedspace** [nushell/nushell#6744]
### Question

I was thinking that `str` subcommands are exclusively commands which take and return strings. However, I've noticed `str index-of` is there, which returns a number. So, shouldn't `split column` and `split chars`, which return structures of strings, be `str` subcommands too? I just want to understand how these commands are batched together.

**Comment by fdncred** [comment]
We could use a good once-over to make our commands more consistent. @dd7 and I were talking about that just yesterday. I always thought `split row` and `split column` were kind of weird too. however, to their benefit, their names do work the way they're named. although, I can never remember if I need to do `split column` or `split row` which says to me, they're confusing still.

**Comment by dandavison** [comment]
I think you're right @webbedspace that, as @fdncred says, command names are not currently named or grouped in a very logical way. It should be a really nice improvement to the language to improve this while the project is still allowing itself to make major changes to the user-facing API.

For this particular issue of `split` and `str`, to make things worse there's also `detect columns` which has a very strange name and in its `-n` form seems to do more or less the same thing as `split column` (just with 0-indexed column names 🤯 ).

**Comment by fdncred** [comment]
I think `detect columns` and `split columns` are very different although I can see from the example why people would think similarly. `detect columns` was written by jt because I was ranting about needed a `columns -t` or `tabulate` (https://github.com/mbudde/tabulate) command that could parse non-structured string data into nushell tables. `detect columns` is more closely related to `from ssv`.

**Comment by dandavison** [comment]
> `detect columns` is more closely related to `from ssv`

I see, thanks. Do you remember why the decision was taken not to expose the functionality under the `from` command, perhaps via a flag on `from ssv`?

**Comment by fdncred** [comment]
I'm just guessing because we have > 400 commands and we forgot about `from ssv`. It seems like `from ssv` really does most of what `detect columns` does, and most of the time better. It's just that the name, `from ssv`, makes you think like `from space separated values` which is similar to `from csv` and one doesn't normally think about parsing the output of ping or other external command line utils with csv-type utilities.

I think the `detect columns` unique code could be integrated into `from ssv` or vice-versa, but we may want to keep both commands and one just work like deprecated commands do, but we just keep it forever. For example, if we just have `from ssv`, we keep `detect columns` and when you run it, it says, "Have you tried from ssv?". I just think the command name `detect columns` is a bit more descriptive and intuitive.

**Comment by webbedspace** [comment]
I personally find `detect columns` to be a rather detestable command name (insofar as `detect` really doesn't make sense for a command that is useful because it does *something* after detecting the columns) and would be glad for its deprecation or concealment.

**Comment by dandavison** [comment]
Thanks @fdncred, I see what `detect columns` does now. It is really useful -- e.g. I used it here: https://github.com/nushell/nu_scripts/tree/main/network/sockets

@webbedspace I agree that "detect columns" doesn't feel like the right name (for one thing, it implies the existence of a family of "detect" commands). I wonder if the functionality could be folded into `from ssv` as @fdncred mentioned.

In a way it has a spiritual connection to awk, which is also used for splitting on whitespace when you don't care how much it has.

I want to mention `parse` here as well for when we're organizing these commands. It looks really helpful and somewhat related.

**Comment by fdncred** [comment]
I think we need to have a balance between having a bunch of commands that do similar things and having one command that has dozens of command line parameters. In general we want to have easy to use intuitive composable commands.

My problem with putting `detect columns` functionality into `from ssv` is really about the name. `from ssv` just doesn't say to me that we want to take some human readable columnar output and transform it to nushell tables. On the other hand, if I have a csv that is delimited by spaces instead of commas, I think of `from ssv`.

Having said that, I could see taking the `detect columns` "guts" and putting it in as `from ssv --detect`. I still have the same problem with the naming but I'm not really sure what to do about it.

Linux has `columns -t` and `tabulate` but I'm not really sure those names are better. I'm open to other suggestions. Anyone have any good ones?

You're right @dandavison, `parse` should be in the same conversation. Maybe we should have `parse columns` that does the `detect columns`/`from ssv` work?

**Comment by dandavison** [comment]
> `from ssv` just doesn't say to me that we want to take some human readable columnar output and transform it to nushell tables.

Right, good point, I can see that! Let's keep iterating on names then. Could it maybe be `from tabular`?

But, while we're really talking about a v2 of nushell builtin names, we might want to consider whether `from xxx` is a good name. Maybe that whole family should be `parse xxx`: `parse csv`, `parse toml`, `parse json`.

Or if we wanted to make it a bit less computer-sciencey it could be `read csv`, `read json`, `read ssv`, etc (which will be familiar to people from Pandas, and R, etc. Then the existing `parse` and `detect columns` could be `read tabular` or `parse tabular`.

**Comment by rgwood** [comment]
My 2 cents on naming:
- `parse` does seem better than `from`, would be a big change but maybe worth it before v1.0
- I'm indifferent about `detect columns` vs `from tabular`. Either is OK IMO

**Comment by webbedspace** [comment]
Something else to consider: if the `from` command is up for renaming, then its verbal counterpart, `to`, should be renamed at the same time.
To me, `to` has the following problems:
* Easily confused with `into` (and by "easily" I mean every time I try to write one, I type the other)
* Even ignoring `into`, easily confused with existing Javascript `toString` and Ruby `to_s`/`to_a` terminology (debatable whether this actually matters)
* `to text` in particular is not clear what exactly `text` means, given they all return strings.
* #6818
It does, however, have these advantages:
* Matches `from`
* Short (even though it isn't exactly common enough to benefit from this shortness)

I think if the "matches `from`" advantage is removed, it should probably be renamed too. It could be folded into `format` subcommands, given that the role of `format` is to convert columnar data to string, and each `to` subcommand could be considered a special `format` case.

**Comment by fdncred** [comment]
FWIW, When I first started learning nushell JT taught me that `into` meant `conversion` and `to` meant output format.

**Comment by dandavison** [comment]
I think:

- We should not do this renaming right now, but it's good to discuss how we think things should be named, including things that already have names.
- We should ship something like #6796 (and maybe changes to input/output type checking internals).
- Then, we will be able to write queries tabulating command names according to their input and output types.
- We should try to agree upon and write down the project's "philosophy" or "naming style" regarding how commands should be named and how the `prefix suffix` subcommand style should be employed (related: discussion on [subcommand delimiter](https://github.com/nushell/nushell/discussions/6460)). Will builtin subcommands continue to be a mixture of `verb noun` and `noun verb`?
- I think that the project should be open to major naming changes. We're competing against all the world's carefully designed programming language standard libraries and pipeline/iterator composition libraries. So far it looks to me like we've mostly allowed individual authors to name things however they feel like, and choose command input/output types and flags however they like. Obviously no-one expects that process to result in a world-class design!
- But we should design it all up-front and then execute on the design. We could possibly introduce a formal "alias command" concept, and go through a period of backward compatibility where the deprecated interface coexists with the new interface, with deprecation warnings.
- Meanwhile we should start requiring peer consensus agreement for user-facing changes to nushell commands.
- Settling on a good naming scheme and always choosing good names (and dividing functionality up appropriately between separate commands) is extremely important to the success of the project.

**Comment by dandavison** [comment]
I agree that `format` is a tempting counterpart to `parse`, for two families of nushell commands that do `structured -> string` and `string -> structured` respectively, and I agree that the naming combination of `to` and `into` is confusing.

To add some more data points in favor of an eventual renaming initiative, I just learned about `grid` (`structured -> string`, seems like another good candidate for the `format` family) and `wrap` (creates a table/list-of-records from iterable input, or a single-key record from non-iterable input; not sure how that fits in but it feels like either there should be a family that it "fits" in, or that it should be idiomatically do-able by using standard table/record constructors and that it shouldn't need its own command).

**Comment by rgwood** [comment]
> I think that the project should be open to major naming changes.

Agreed.

**Comment by webbedspace** [comment]
As an addendum to the original post: I think `split rows`, which is the "fundamental" split analogous to JS and Python `.split()`, is too difficult to discover and should not have `rows` at all. I think the `columns` version is also pretty bizarre, insofar as one can simply add `| rotate -ccw` to `rows`'s output to get the same result.

**Comment by theAkito** [comment]
I just fiddled around with `detect columns -n` & `split column`.

Indeed. It is very confusing. Had to try out both commands thoroughly to get how they actually are working.

For example, I expected `split column` to split columns, similarly to `detect columns -n`. However, to be able to do that, you need to pipe the input through `lines`, before, in my specific scenario.
Anyway, it's confusing.

Here are several ways of achieving the same thing: getting all usernames from `/etc/passwd`.

```nu
open --raw /etc/passwd | str replace --all --string ':' (char space) | detect columns -n | get column0
open --raw /etc/passwd | lines | par-each { |it| $it | split column ':' name | get 0.name }
open --raw /etc/passwd | lines | split column ':' | get column1
```

**Comment by fdncred** [comment]
This is how I would parse the /etc/passwd on my Mac, not sure it's the same on Linux.
```
 open /etc/passwd | lines | skip 10 | parse "{user}:{pass}:{userid}:{groupid}:{fullname}:{path}:{other}"
```
or alternately
```
open /etc/passwd | lines | skip 10 | to text | from csv --separator ':'
```

`detect columns` and `from ssv` are really for parsing space delimited files.

I'm not sure exactly how to use `split columns` myself, other than what the examples show.

**Comment by dead10ck** [comment]
It's been a little while since this was reported. Has there been any further consideration, given all the changes that have happened? I still trip up on this, typing `str split` and get confused when it doesn't work

**Comment by fdncred** [comment]
I figured we'll keep it as-is since there are so many split commands right now.
```nushell
Subcommands:
  split cell-path - Split a cell-path into its components.
  split chars - Split a string into a list of characters.
  split column - Split a string into multiple columns using a separator.
  split list - Split a list into multiple lists using a separator.
  split row - Split a string into multiple rows using a separator.
  split words - Split a string's words into separate rows.
```
I could be persuaded otherwise. People could also probably use aliases like `alias 'str split-row' = split row`

**Comment by WindSoilder** [comment]
Going to close the issue because it's answered.
